### ISSUE TITLE
Better control of minimap flipping

### ISSUE THREAD

**Issue by WatchTheFort** [beyond-all-reason/RecoilEngine#618]
The config int `MiniMapCanFlip` controls whether or not the minimap automatically rotates 180° when the camera is rotated past the plane of 180°. This is wanted behaviour when using the camera flip command (the Ctrl-Shift-O `cameraflip` action in the game), but is unwanted when manually rotating the camera, due to breaking the cinematography 180° rule.

Is there a command available to manually flip the minimap, which could be invoked when calling `cameraflip`, but when `MiniMapCanFlip` is 0?

**Comment by WatchTheFort** [comment]
@badosu

**Comment by sprunk** [comment]
No engine change needed, the engine already exposes the tools. Do `Spring.SetConfigInt("MiniMapCanFlip", 0)` when enabling a free-rotation camera and `Spring.SetConfigInt("MiniMapCanFlip", 1)` when enabling the regular camera that can only flip.

**Comment by badosu** [comment]
Yes, pretty much what @sprunk said. If the camera changes to whatever you don't want the minimap to flip on (say Spring camera), then disable `MiniMapCanFlip`.

Additionally, I think it makes sense to have a Callin for when camera controller changes, thoughts?

**Comment by WatchTheFort** [comment]
There shouldn't even be camera changes because there should only be one camera, but that's unrelated to this issue.

**Comment by badosu** [comment]
I don't necessarily disagree with that in some sense, but we gotta deal with things in retrocompatible and smooth ways.

**Comment by badosu** [comment]
I can expose `Spring.SetMiniMapRotation`, which was planned for the general behavior and gracefully enhances for the flipping behavior. Does that address the issue for you?

**Comment by WatchTheFort** [comment]
That sounds workable, for now I'll implements Sprung's suggestion, since I've broken it for players who play with a flipped map.

**Comment by badosu** [comment]
I dislike the current way `DrawInMiniMap` happens to be honest. I'm not at all comfortable with the engine not handling minimap rotation transformations itself.

It does not make sense for the game developers to have to apply those transformations themselves, and they can always get true north with `GetMiniMapRotation`. Thoughts @sprunk?

**Comment by WatchTheFort** [comment]
@sprunk Unfortunately, it didn't work, the minimap does not flip when the map flips.
```
local miniMapCanFlip = Spring.GetConfigInt('MiniMapCanFlip', 0)

if camState.flipped then
	Spring.SetConfigInt('MiniMapCanFlip', 1)
	camState.flipped = camState.flipped * -1
	Spring.SetCameraState(camState, 0)
	Spring.SetConfigInt('MiniMapCanFlip', miniMapCanFlip)
end
```
I checked, and it's hitting everything inside there.

**Comment by badosu** [comment]
For now you can mitigate for players with `/Set MiniMapCanFlip 1`

And in the place where you hardcode it to 0, just don't hardcode it (default is 0).

**Comment by badosu** [comment]
For the players that didn't get the change to 0, `Set MiniMapCanFlip 0`.

Arguably I don't get "the cinematography rule" for gameplay, if I have free rotation I prefer to have the minimap in the orientation that aligns with my north, although I've heard some players say they dislike the flipping.

For casting in particular, there should be a setting for "Casting Mode" that sets sane defaults for casting specifically.

**Comment by WatchTheFort** [comment]
It's cinematography not as in movies, but as in a camera and a viewer. The issue is that it flips 0°-180°-0°, reversing left and right, causing the viewer to lose their orientation and points of reference. This is different than having a the minimap use continuous rotation.

**Comment by sprunk** [comment]
```
local miniMapCanFlip = Spring.GetConfigInt('MiniMapCanFlip', 0)

if camState.flipped then
	Spring.SetConfigInt('MiniMapCanFlip', 1)
	camState.flipped = camState.flipped * -1
	Spring.SetCameraState(camState, 0)
	Spring.SetConfigInt('MiniMapCanFlip', miniMapCanFlip)
end
```
Why are you relying on `flipped`, isn't that only for the regular overhead camera and doesn't apply to the cameras that rotate?
Why are you setting the config int back to the original value? That sounds like it would revert whatever you did in SetCameraState.

I thought you'd just do something like this https://github.com/beyond-all-reason/Beyond-All-Reason/commit/79a7803f47b67c625d71dfb5416bfcd6a52faa71

**Comment by sprunk** [comment]
I've learnt that a shader can be applied to the minimap so a game can apply whatever it wants to the rendering, incl. flipping. I'm not closing yet though because there is no good prior art so there are likely to be issues (e.g. what about giving orders via rightclick?).

**Comment by sprunk** [comment]
9d2c24d68180fee2ab50a80a211ee9fcf8ef27b2 adds `Spring.SetMiniMapRotation`.
